src/keyboard/mod.rs

```diff
@@ -569,59 +604,267 @@ where
+                                                keysym: thread_sym,
+                                                utf8: thread_utf8.clone(),
+                                            },
+                                            proxy.clone(),
+                                        );
+                                        // Rate
+                                        thread::sleep(Duration::from_millis(repeat_timing.0));
+                                        match thread_kill_chan.lock().unwrap().1.try_recv() {
+                                            Ok(_) | Err(mpsc::TryRecvError::Disconnected) => {
+                                                break
+                                            }
+                                            _ => {}
+                                        }
+                                    }
+                                });
                             }
-                            ffi::xkb_compose_status::XKB_COMPOSE_NOTHING => state.get_utf8_raw(key),
-                            _ => None,
                         }
                     } else {
-                        state.get_utf8_raw(key)
-                    };
-                    let modifiers = state.mods_state.clone();
-                    user_impl.receive(
-                        Event::Key {
-                            serial,
-                            time,
-                            modifiers,
-                            rawkey: key,
-                            keysym: sym,
-                            state: key_state,
-                            utf8,
-                        },
-                        proxy,
-                    );
+                        if key_held == Some(key) {
+                            kill_chan.lock().unwrap().0.send(()).unwrap();
+                            key_held = None;
+                        }
+                        event_impl.receive(
+                            Event::Key {
+                                serial,
+                                time,
+                                modifiers,
+                                rawkey: key,
+                                keysym: sym,
+                                state: key_state,
+                                utf8: utf8.clone(),
+                            },
+                            proxy.clone(),
+                        );
+                    }
                 }
                 wl_keyboard::Event::Modifiers {
                     mods_depressed,
                     mods_latched,
                     mods_locked,
                     group,
                     ..
-                } => state.update_modifiers(mods_depressed, mods_latched, mods_locked, group),
+                } => {
+                    state.lock().unwrap().update_modifiers(mods_depressed, mods_latched, mods_locked, group);
+                    state_chan.lock().unwrap().0.send(()).unwrap();
```

Comment:
Doesn't that stop any current repetition if one was going on?
Pressing a modifier should change the repeated key but not stop the repetition.

Also, what's the impact of sending a message in this channel if no repetition is going on?